server/src/main/java/org/elasticsearch/search/aggregations/metrics/TopHitsAggregatorFactory.java

```diff
@@ -118,8 +124,41 @@ public Aggregator createInternal(SearchContext searchContext,
         if (highlightBuilder != null) {
             subSearchContext.highlight(highlightBuilder.build(searchContext.getQueryShardContext()));
         }
+        if (isNested(parent) && searchContext.parsedQuery() != null) {
```

Comment:
I am not sure we should do that, we have the same problem on clauses that appear under a `nested` query. If the fetch phase is run inside `inner_hits` we'll break the matching. I am in favor of documenting this behavior and maybe add an option to consider each named query as independent like we do today. I think it's an improvement to consider that a named query matches only if the other required clauses are matching but we cannot handle multi-level easily.